feat: add verify option for transmission-show

srussel opened the pull request:
Fixes #6089

ckerr commented on libtransmission/verify.cc:
```diff
@@ -118,24 +153,43 @@ bool tr_verify_worker::verify_torrent(tr_torrent* tor, std::atomic<bool> const&
+                    if (offline_results)
+                    {
+                        result.fail = true;
+                        std::string error = fmt::format("Piece {:d} hash fail", piece);
```
```suggestion
                        auto const error = fmt::format("Piece {:d} hash fail", piece);
```

ckerr commented on libtransmission/transmission.h:
```diff
@@ -1397,6 +1405,22 @@ void tr_torrentAmountFinished(tr_torrent const* torrent, float* tab, int n_tabs)
 
+#ifdef __cplusplus
+
+struct tr_verify_file_result
+{
+    std::string_view filename;
+    uint64_t size;
+    std::vector<std::string> errors;
+    bool fail;
+};
```
:thinking: 

Instead of having this in the public API, what would you think about `tr_verify_worker::verify_torrent()` taking a third argument that is an error callback, e.g.

```c++
using tr_verify_error_cb = void(*)(
    std::string_view filename,
    uint64_t size,
    std::string_view error);

...

[[nodiscard]] bool tr_torrentImmediateVerify(tr_torrent* torrent, tr_verify_error_cb cb = nullptr);
```

IMO this would be more flexible and have a smaller public API footprint

srussel replied:
I was on the fence with using a callback vs passing a vector when I wrote the code. Using a vector seemed to be easier. But I'm happy to change it to a callback. An advantage of using the callback is if `--unsorted` is used, we can output the status immediately instead of accumulating it and then outputting it at the end. This is useful with huge torrents where you could be waiting a while before you see the files listed.

ckerr commented on utils/transmission-show.1:
```diff
@@ -19,11 +19,39 @@ shows BitTorrent .torrent file metadata
 .Bl -tag -width Ds
 .It Fl h Fl -help
 Show a short help page and exit.
+.It Fl d Fl -header
+Show only header section.
+.It Fl i Fl -info
+Show only info section.
+.It Fl t Fl -trackers
+Show only trackers section.
+.It Fl f Fl -files
+Show only header section.
+.It Fl D Fl -no-header
+Do not show header section.
+.It Fl I Fl -no-info
+Do not show info section.
+.It Fl T Fl -no-trackers
+Do not show trackers section.
+.It Fl F Fl -no-files
+Do not show files section.
+.It Fl b Fl -bytes
+Show file sizes in bytes.
```
:heart:

ckerr commented:
@srussel To make this PR easier, I've added a new PR at https://github.com/transmission/transmission/pull/6123 that changes `tr_verify_worker` to no longer require a `tr_torrent` in order to work.

srussel commented:
> @srussel To make this PR easier, I've added a new PR at #6123 that changes `tr_verify_worker` to no longer require a `tr_torrent` in order to work.

That would be great. I'm unfamiliar with the transmission codebase and the change I did to get verify to work without a session felt like a hack.

What is the preference for updating my change after #6123 lands? Merge main and keep adding commits? Or rebase and force push?

ckerr commented:
> What is the preference for updating my change after 6123 lands? Merge main and keep adding commits? Or rebase and force push?

I generally prefer rebase+force-push but TBH whatever's easier for you is fine. Generally it's all effectively the same in the end since PRs like 6112 and 6123 each get squash-merged into `main` as a single commit.

tearfur commented:
I'll take a look once #6123 lands.

ckerr commented:
@srussel I've merged 6123

srussel commented:
I've updated the PR with the following changes
- use the mediator interface in verify worker
- use a callback in the verify function
- in show.cc, if unsorted, print the file status as files are checked instead of waiting until the end.

srussel commented:
Fixed a clang-tidy warning.

nevack commented:
@ckerr Can you approve running workflows/actions here?

srussel commented:
More clang-tidy fixes.

tearfur commented on utils/show.cc:
```diff
@@ -468,11 +599,28 @@ int tr_main(int argc, char* argv[])
+                /* If opts.verify, verify() will print the file list including the status.
+                 * So disable the file list in showInfo(). */
+                print_ok_files = true;
+                opts.print_files = false;
```
Looks to me that `opts.print_files` should be set inside `parseCommandLine()` to be more consistent with the existing code organisation.

Then you can call `verify()` like `verify(opts, opts.print_files && opts.verify, metainfo)`.

tearfur replied:
Or, you can change

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/utils/show.cc#L270

to

```cpp
if (opts.print_files && !opts.verify)
```

So that
```cpp
bool verify(app_opts const& opts, bool print_ok_files, tr_torrent_metainfo const& metainfo)
```
can directly check `opts.print_file` instead of `print_ok_files` and become
```cpp
bool verify(app_opts const& opts, tr_torrent_metainfo const& metainfo)
```
instead.

tearfur commented on utils/show.cc:
```diff
@@ -468,11 +599,28 @@ int tr_main(int argc, char* argv[])
+            }
             showInfo(opts, metainfo);
+
+            if (opts.verify)
+            {
+                bool ok = verify(opts, print_ok_files, metainfo);
+                if (!ok)
+                {
+                    exit_status = EXIT_FAILURE;
+                }
+            }
```
To avoid deeply-nested code, looks like this block can be moved up to the same level as `if (opts.print_header)` and `if (opts.scrape)`.

srussel replied:
I've moved it but I am wondering why `showInfo()` is not called when `opts.scrape` is enabled.

tearfur replied:
I don't see why not. If you're interested, you're welcome to submit another PR for it :)

tearfur commented on libtransmission/verify.cc:
What is the reason for removing the hard-coded 100ms break here?

Judging from this comment, it sounds to me sleeping every once in a while is always a good idea:

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L145-L146

tearfur commented on libtransmission/verify.cc:
```diff
@@ -83,23 +102,59 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
         /* if we're finishing a piece... */
         if (left_in_piece == 0)
         {
             auto const has_piece = sha->finish() == metainfo.piece_hash(piece);
             verify_mediator.on_piece_checked(piece, has_piece);
+            if (has_piece)
+            {
+                for (auto const file : files_ending_in_current_piece)
+                {
+                    verify_mediator.on_file_ok(file);
+                }
+            }
+            else
+            {
+                auto const error{ fmt::format("Piece {:d} hash fail", piece) };
+                for (auto const file : files_ending_in_current_piece)
+                {
+                    verify_mediator.on_file_error(file, error);
+                }
+                if (left_in_file != 0)
+                {
+                    verify_mediator.on_file_error(file_index, error);
+                }
+                file_valid = false;
+            }
 
-            /* sleeping even just a few msec per second goes a long
-             * way towards reducing IO load... */
-            if (auto const now = current_time_secs(); last_slept_at != now)
+            if (sleep_ms_per_second > 0ms)
             {
-                last_slept_at = now;
-                std::this_thread::sleep_for(SleepPerSecondDuringVerify);
+                /* sleeping even just a few msec per second goes a long
+                 * way towards reducing IO load... */
+                if (auto const now = current_time_secs(); last_slept_at != now)
+                {
+                    last_slept_at = now;
+                    std::this_thread::sleep_for(SleepPerSecondDuringVerify);
+                }
```
Maybe you can enlighten me about the reasonings for allowing to disable the 100ms breaks.
Judging from this comment, it sounds to me sleeping every once in a while is always a good idea:

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L145-L146

If we all agree in the end that allowing to disable the 100ms breaks is a good idea, then `sleep_ms_per_second` should either be replaced with a `bool` parameter, or be passed as a parameter to `sleep_for()`. The current code here looks weird.

srussel replied:
My understanding of that comment is that when running a session with other torrents downloading, you want to ensure the verify does not slow down the other torrents. The change I am adding is for a command line utility that runs in a single thread with no other IO. In this case the user expects the verify to run as quick as possible, similar to running other IO intensive utils like `cp` or `tar`.

I intended to pass `sleep_ms_per_second` to `sleep_for()`. That was a mistake.

tearfur replied:
> In this case the user expects the verify to run as quick as possible

True, but as a user, I'd appreciate it if Transmission does not starve my other IO-intensive applications, or cause an overall slowdown in any way. So I think it's better to revert this change.

srussel replied:
You can use `ionice` if you want to limit IO. I would prefer not to slow it down when I just want it to run as fast as possible. Could be add a command line option to limit the IO?

GaryElshaw replied:
Thoughts Charles?

tearfur replied:
Hmm, looks like Linux can use [ioprio_set](https://man7.org/linux/man-pages/man2/ioprio_set.2.html) and Windows can use [SetProcessInformation](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessinformation).

ckerr replied:
> Thoughts Charles?

IMO our `std::this_thread::sleep_for()` call is a little crude, but it's a one-liner that is working fine, is portable code, needs no error-handling and no ifdefs. I'd keep it as-is unless there's a known issue with it.

> My understanding of that comment is that when running a session with other torrents downloading, you want to ensure the verify does not slow down the other torrents.

`tr_verify_worker` runs in its own thread, so there shouldn't be any issue of verify blocking the session thread.

tearfur commented on libtransmission/torrent.h:
```diff
@@ -97,6 +97,12 @@ struct tr_torrent final : public tr_completion::torrent_view
         void on_verify_started() override;
         void on_piece_checked(tr_piece_index_t piece, bool has_piece) override;
         void on_verify_done(bool aborted) override;
+        void on_file_ok(tr_file_index_t) override
+        {
+        }
+        void on_file_error(tr_file_index_t, std::string) override
+        {
+        }
```
clang-tidy will complain about these parameters having no name.

You can do something like `tr_file_index_t /*index*/` to avoid the warnings.

tearfur commented on libtransmission/verify.h:
```diff
@@ -37,6 +44,8 @@ class tr_verify_worker
         virtual void on_piece_checked(tr_piece_index_t piece, bool has_piece) = 0;
         virtual void on_verify_done(bool aborted) = 0;
+        virtual void on_file_ok(tr_file_index_t index) = 0;
+        virtual void on_file_error(tr_file_index_t index, std::string error) = 0;
```
```suggestion
        virtual void on_file_error(tr_file_index_t index, std::string_view error) = 0;
```

You avoid copying the error string in this way.

tearfur reviewed (commented):
This is just a partial review, I will continue when I've got more free time.

tearfur commented on utils/show.cc:
```diff
@@ -399,11 +405,136 @@ void doScrape(tr_torrent_metainfo const& metainfo)
+
+void verify_callback(tr_file_index_t index, bool ok, std::string error)
+{
+    if (index != current_index)
```
Please correct me if I am wrong here.

AFAICT, `verify_callback()` is called at most once for each file, so this if condition isn't needed. In fact, this is causing the last file to *not* be printed, which you worked around by repeating code inside `verify()`, so there's a very good reason to remove it.

Following the same logic, we can also do away with `verify_cb::errors` because each file will have at most 1 error getting printed out. `verify_callback()` can probably be drastically simplified after this change.

srussel replied:
When there are errors `verify_callback()` can be called multiple times per file. eg Multiple failed pieces in the one file. I could make it one callback per file with a vector of errors but it makes the `verify_torrent()` logic more complicated. It doesn't help that a piece can contain multiple files.

tearfur replied:
> When there are errors `verify_callback()` can be called multiple times per file. eg Multiple failed pieces in the one file. 

Does it? I thought the error callback will only be called for the first piece that errored, because `file_valid` will be set to false here:

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L140

And so the file won't be added to `files_ending_in_current_piece` for the pieces afterwards.

> I could make it one callback per file with a vector of errors but it makes the `verify_torrent()` logic more complicated. It doesn't help that a piece can contain multiple files.

No no, that's not what I want either. In that case, how about you initialise `current_index` to `~0` instead of `0`, and set `status.index` to `index` instead of `current_index` inside `verify_callback()`?

```cpp
// Before
tr_file_index_t current_index;
// After
tr_file_index_t current_index = ~0;

// Before
status.index = current_index;
// After
status.index = index;
```

tearfur replied:
> When there are errors `verify_callback()` can be called multiple times per file. eg Multiple failed pieces in the one file. 

Oh! I see it now. Yes you are right.

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L136-L139

srussel replied:
I was thinking that if any piece fails we could just report "hash fail" for the affected file(s) instead of reporting each piece that failed. This would simplify the code as there would only be one callback per file - either success, some type of file error (unable to open, wrong size, etc), or hash fail. I could merge `on_file_ok` and `on_file_error` into one callback.

I don't think it makes much difference to the end user whether we report only that a file failed the hash check or report every piece that failed. Either way the user can't trust the file.

tearfur replied:
I think either way is acceptable, feel free to write whatever that's comfortable for you.

I'd say keep the current behaviour though, since you've already put in the effort, and the fix for the repeated code seems to be very simple:

> In that case, how about you initialise `current_index` to `~0` instead of `0`, and set `status.index` to `index` instead of `current_index` inside `verify_callback()`?

tearfur commented on utils/show.cc:
```diff
@@ -399,11 +405,136 @@ void doScrape(tr_torrent_metainfo const& metainfo)
+
+    std::string_view data_dir{ opts.verify };
+    bool all_ok = tr_torrentSynchronousVerify(&metainfo, &data_dir, verify_cb::verify_callback);
+    if (!verify_cb::errors.empty())
+    {
+        /* Last file has errors. Print or store the status as required. */
+        verify_cb::Status status;
+        status.index = verify_cb::current_index;
+        status.ok = false;
+        status.errors = verify_cb::errors;
+        if (opts.unsorted)
+        {
+            verify_cb::print_file(status);
+        }
+        else
+        {
+            verify_cb::files.emplace_back(status);
+        }
+    }
```
This is repeating code from `verify_callback()`.

tearfur reviewed (changes requested):
@srussel I think that's all the comments I have for now. Thank you for writing this!

srussel commented on libtransmission/verify.cc:
```diff
@@ -54,8 +56,25 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
         {
+            file_valid = true;
             auto const found = verify_mediator.find_file(file_index);
             fd = !found ? TR_BAD_SYS_FILE : tr_sys_file_open(found->c_str(), TR_SYS_FILE_READ | TR_SYS_FILE_SEQUENTIAL, 0);
+            if (fd == TR_BAD_SYS_FILE)
+            {
+                file_valid = false;
+                verify_mediator.on_file_error(file_index, std::string("Missing file."));
```
That is a good idea. My intention is to report the errors as accurately as possible.

tearfur replied:
@srussel Just a heads up, the `tr_error` API is going to change with #6198.

srussel commented:
The remaining comments look straightforward and I will work on them sometime in the next few days.

ckerr commented on libtransmission/verify.cc:
```diff
@@ -45,6 +45,8 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
     auto sha = tr_sha1::create();
     auto last_slept_at = current_time_secs();
+    bool file_valid = false;
+    std::vector<tr_file_index_t> files_ending_in_current_piece;
```
It feels like the piece-to-file housekeeping here and in verify_callback() are repeating some code that's already written + unit tested in `file-piece-map.h` but is currently hidden behind a `#error only libtransmission should #include this header.` compile guard. 

@srussel WWYT if `tr_torrent_metainfo` had the `file_piece_map::file_span()` function so that you could get a range of files associated with a piece? Would we even still need `files_ending_in_current_piece` or `on_file_ok()`?

ckerr replied:
@srussel ping

srussel replied:
I do intend to finish this but I won't have time to work on it until mid-December.

srussel replied:
> @srussel what would you think if `tr_torrent_metainfo` had the `file_piece_map::file_span()` function so that you could get a range of files associated with a piece? Would we even still need `files_ending_in_current_piece` or `on_file_ok()`?

I have not looked into this yet. I wanted to push an update with all the other issues addressed first before I do any significant changes to the verify function so I have something working that I can compare it to.

tearfur replied:
> I have not looked into this yet. I wanted to push an update with all the other issues addressed first before I do any significant changes to the verify function so I have something working that I can compare it to.

@srussel Is this still on-going?

srussel replied:
No, I've been busy again with other things. Is this something you need me to do in order to merge this PR? It's just that it could be a while before I get the time to work on it.

ckerr commented on libtransmission/verify.cc:
```diff
@@ -83,23 +102,59 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
+             */
+            if (file_valid)
+            {
+                files_ending_in_current_piece.push_back(file_index);
```
if file X _begins_ in the current piece but doesn't end in the piece, and the piece checksum fails, will `on_file_error()` be called for file X?

tearfur replied:
As per my previous discussion with the author, it will here:

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L136-L139

tearfur commented on libtransmission/torrent.h:
```diff
@@ -97,6 +97,12 @@ struct tr_torrent final : public tr_completion::torrent_view
+        void on_file_ok(tr_file_index_t) override
+        {
+        }
+        void on_file_error(tr_file_index_t, std::string) override
```
I already pointed this out at https://github.com/transmission/transmission/pull/6112#discussion_r1373295187.

srussel commented:
I'm getting back up to speed on this. I've rebased and pushed the changes I already have. I think that addresses almost everything.

srussel commented:
Github won't let me reply to the review thread on the sleep() probably due to the code changing so I have quoted and added my response below.

>>    Thoughts Charles?

>IMO our std::this_thread::sleep_for() call is a little crude, but it's a one-liner that is working fine, is portable code, needs no error-handling and no ifdefs. I'd keep it as-is unless there's a known issue with it.

>>    My understanding of that comment is that when running a session with other torrents downloading, you want to ensure the verify does not slow down the other torrents.

>tr_verify_worker runs in its own thread, so there shouldn't be any issue of verify blocking the session thread.

I've simplified my change a bit so it is just a bool flag to disable the sleep.

Just to be clear, I have no issue with the verify worker thread sleeping and have left this as the default. The issue I have is when using `transmission-show`, I am interactively using the command line. I run `transmission-show --verify ...` then sit there staring at the screen waiting for the command to finish. In this case sleeping for 100ms per second is an unacceptable delay. It makes the verify take 10% longer for no reason. The change to disable the sleep only occurs when using the  `tr_torrentSynchronousVerify` function, which does not use the worker thread, and is only used by `transmission-show`. There is no change to the behavior when using verify worker in session.cc.

andreygursky commented on libtransmission/verify.cc:
```diff
@@ -84,23 +117,59 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
-            /* sleeping even just a few msec per second goes a long
-             * way towards reducing IO load... */
-            if (auto const now = current_time_secs(); last_slept_at != now)
+            if (limit_io)
```
See also #6572.

Coeur commented:
@srussel can you rebase to address the conflicts? Thank you

srussel commented:
@Coeur I've rebased this PR
